src/graph.rs

```diff
@@ -86,6 +92,131 @@ impl<'tree> Graph<'tree> {
 
         DisplayGraph(self, ctx)
     }
+
+    pub fn display_json<'a>(&'a self, ctx: &'a Context) -> () {
+        // TODO: move this all into a new module
+        struct JSONGraph<'a, 'tree>(&'a Graph<'tree>, &'a Context);
+        struct JSONNode<'a>(usize, &'a GraphNode, &'a Context);
+        struct JSONEdge<'a>(&'a u32, &'a Edge, &'a Context);
+        struct JSONAttributes<'a>(&'a Attributes, &'a Context);
+        struct JSONIdentifier<'a>(&'a Identifier, &'a Context);
+        struct JSONValue<'a>(&'a Value, &'a Context);
+
+        impl<'a, 'tree> ser::Serialize for JSONGraph<'a, 'tree> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let graph = self.0;
+                let ctx = self.1;
+                let mut seq = serializer.serialize_seq(Some(graph.graph_nodes.len()))?;
+                for (node_index, node) in graph.graph_nodes.iter().enumerate() {
+                    seq.serialize_element(&JSONNode(node_index, node, ctx))?;
+                }
+                seq.end()
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONNode<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let node_index = self.0;
+                let node = self.1;
+                let ctx = self.2;
+                // serializing as a map instead of a struct so we don't have to encode a struct name
+                let mut map = serializer.serialize_map(None)?;
+                map.serialize_entry("index", &node_index)?;
+                if !node.outgoing_edges.is_empty() {
+                    let edges = Vec::from_iter(
+                        node.outgoing_edges
+                            .iter()
+                            .map(|(sink, edge)| JSONEdge(sink, edge, ctx)),
+                    );
+                    map.serialize_entry("edges", &edges)?;
```

Comment:
It was this or make another inner struct just for vectors of edges. Tragically, there is no `Serialize` implementation for `SmallVec`.

Note that `collect_seq` is not applicable since `serialize` returns a `Result` of unit and not, say, something which you could pass in as a `Serialize`-implementing value for the second argument to `serialize_entry` here. (I take issue with several points of `serde`'s design.)